### ISSUE TITLE
Initializing conda shell integration on self-hosted runner

### ISSUE THREAD

**Issue by martin-g** [conda-incubator/setup-miniconda#205]
This issue is a continuation of https://github.com/conda-incubator/setup-miniconda/issues/203.
My reasons to create a new issue are:
1) to stop abusing the other issue - it is somehow related but it is also closed 
2) now I have some progress with running setup-miniconda on self-hosted runner and maybe the new issue is a bug 

Here I was able to use setup-miniconda in a self-hosted runner:
1) I've created a custom script that setups Conda:
```
#!/usr/bin/env bash


export CONDA="/home/ubuntu/miniconda/"
export PATH="$CONDA/condabin:$PATH"

# >>> conda initialize >>>
# !! Contents within this block are managed by 'conda init' !!
__conda_setup="$('/home/ubuntu/miniconda/bin/conda' 'shell.bash' 'hook' 2> /dev/null)"
if [ $? -eq 0 ]; then
    eval "$__conda_setup"
else
    if [ -f "/home/ubuntu/miniconda/etc/profile.d/conda.sh" ]; then
        . "/home/ubuntu/miniconda/etc/profile.d/conda.sh"
    else
        export PATH="/home/ubuntu/miniconda/bin:$PATH"
    fi
fi
unset __conda_setup
# <<< conda initialize <<<
```
2) `source`d it at the top of the runner's runsvc.sh (this is the script called by systemd's `ExecStart=...`)

The new problem is: https://github.com/martin-g/zeppelin/runs/4532009932?check_suite_focus=true

```
Initializing conda shell integration...
  Removing "/home/ubuntu/.bashrc"
  Removing "/home/ubuntu/.profile"
  /home/ubuntu/miniconda/condabin/conda init --all
  no change     /home/ubuntu/miniconda/condabin/conda
  no change     /home/ubuntu/miniconda/bin/conda
  no change     /home/ubuntu/miniconda/bin/conda-env
  no change     /home/ubuntu/miniconda/bin/activate
  no change     /home/ubuntu/miniconda/bin/deactivate
  no change     /home/ubuntu/miniconda/etc/profile.d/conda.sh
  no change     /home/ubuntu/miniconda/etc/fish/conf.d/conda.fish
  no change     /home/ubuntu/miniconda/shell/condabin/Conda.psm1
  modified      /home/ubuntu/miniconda/shell/condabin/conda-hook.ps1
  no change     /home/ubuntu/miniconda/lib/python3.8/site-packages/xontrib/conda.xsh
  no change     /home/ubuntu/miniconda/etc/profile.d/conda.csh
  modified      /home/ubuntu/.bashrc
  modified      /home/ubuntu/.zshrc
  modified      /home/ubuntu/.config/fish/config.fish
  modified      /home/ubuntu/.xonshrc
  modified      /home/ubuntu/.tcshrc
  
  ==> For changes to take effect, close and re-open your current shell. <==
  
  Renaming "/home/ubuntu/.bashrc" to "/home/ubuntu/.profile"
  
Error: ENOENT: no such file or directory, rename '/home/ubuntu/.bashrc' -> '/home/ubuntu/.profile'
```

It seems to me that setup-miniconda removes ~/.bashrc and ~/.profile and later complains that they are not existing.

**Comment by martin-g** [comment]
OK. I think I've found the problem.
All is good after removing everything between `# >>> conda initialize >>>` and `# <<< conda initialize <<<` and setup-miniconda do all this!

**Comment by jaimergp** [comment]
Oh, nice! Yes, I guess once shouldn't initialize the local installation and let the action do it for them instead. Thanks for debugging!

**Comment by martin-g** [comment]
One question:

On Github runners (x86_64) a project uses:
```
 - name: Setup conda environment with python ${{ matrix.python }} and R
        uses: conda-incubator/setup-miniconda@v2
        with:
          python-version: 3.7
          ...
```

But on my self-hosted runner I installed Miniconda myself with 
```
wget https://repo.anaconda.com/miniconda/Miniconda3-py38_4.10.3-Linux-aarch64.sh -O miniconda.sh
bash miniconda.sh -b -p $HOME/miniconda
```
i.e. Miniconda + Python 3.8.

Do I need to install a second Miniconda3 with Python 3.7 in another folder and use runner labels to decide on which runner to run the job depending on ${matrix.python} ?
Or by using:
```
with:
          python-version: 3.7
```

it will do some magic and work as desired ?

**Comment by jaimergp** [comment]
As long a conda installation is in the system, the action will create envs / adapt python versions as needed. I don't know how concurrency is handled though, I get a feeling that the action is written with the assumption of "each machine running is a VM that will get cleared once we are done" so it's not clear what will happen if two jobs are using the same self-hosted machine with no VM abstraction (don't know if that's your case).

**Comment by martin-g** [comment]
> As long a conda installation is in the system, the action will create envs / adapt python versions as needed.

That's good!

> two jobs are using the same self-hosted machine with no VM abstraction (don't know if that's your case).

This is the case here!
Self-hosted runners are not disposable VMs. Github Actions just guarantees that a single runner always runs at most one job. Once the job is done the runner might be used as is for another job.
I have 6 runners on the same VM, so I will see soon whether collisions will break something.

Is there a way to clean up Miniconda related stuff at the end of a job ?

**Comment by jaimergp** [comment]
You can have the action install a Miniconda of your choosing via `installer-url` (or similar, check the README). Then maybe you can add a cleanup step (or maybe the action cleans after itself?) to remove the directory. Adding a job identifier to the installation path might help too?

**Comment by martin-g** [comment]
It is a bit more complicated - the setup works fine for `ubuntu-20.04` (i.e. Github hosted runner). I am trying to add testing on Linux ARM64 to the setup (hence the self-hosted runners).
If I add `installer-url` then it will affect the x86_64 runners too.
I try to avoid making changes in the job definitions. The best would be if the only change in the workflow YAML is in `runs-on`.

setup-miniconda modifies `~/.bashrc`/`~/.profile`, right ? It seems this causes problems on the self-hosted runner (one VM with several runners) - each job overrides these files
